packages/assets-controllers/src/selectors.ts

```diff
@@ -0,0 +1,100 @@
+import { convertHexToDecimal } from '@metamask/controller-utils';
+import type { Hex } from '@metamask/utils';
+import { createSelector } from 'reselect';
+
+import type { CurrencyRateState } from './CurrencyRateController';
+import type { TokenBalancesControllerState } from './TokenBalancesController';
+import type { TokenRatesControllerState } from './TokenRatesController';
+import type { TokensControllerState } from './TokensController';
+
+const currencyConversionRate = (state: CurrencyRateState) =>
+  state.currencyRates;
+
+const evmTokenBalances = (state: TokenBalancesControllerState) =>
+  state.tokenBalances;
+
+const evmTokenRates = (state: TokenRatesControllerState) => state.marketData;
+
+const tokensInformations = (state: TokensControllerState) => state.allTokens;
+
+export const evmAggregatedBalancesForAccount = createSelector(
+  evmTokenBalances,
+  evmTokenRates,
+  currencyConversionRate,
+  tokensInformations,
+  (_state: TokenBalancesControllerState, account: Hex) => account,
+  (
+    tokenBalances,
+    tokenRates,
+    currencyConversionRateState,
+    tokensInformationsState,
+    account,
+  ) => {
+    const balancesForAccount = tokenBalances?.[account];
+    if (!balancesForAccount) {
+      return {};
+    }
+
+    // Calculate total balance value across all chains for this account
+    let totalBalance = 0;
+
+    for (const [chainId, chainBalances] of Object.entries(balancesForAccount)) {
+      for (const [tokenAddress, balance] of Object.entries(chainBalances)) {
+        // tokenRates is organized by chainId -> tokenAddress -> MarketDataDetails
+        // We need to find the market data for this token across all chains
+        let marketData;
+        for (const chainData of Object.values(tokenRates)) {
+          if (chainData[tokenAddress as Hex]) {
+            marketData = chainData[tokenAddress as Hex];
+            break;
+          }
+        }
+
+        const rateToUse =
+          currencyConversionRateState[marketData?.currency ?? '']
+            ?.conversionRate ?? 0;
+
+        if (marketData?.price) {
+          const { price } = marketData;
+
+          // Find the token decimals from tokensInformations
+          let tokenDecimals = 18; // Default to 18 decimals
+          const chainTokens = tokensInformationsState[chainId as Hex];
+          if (chainTokens) {
+            // Look through all accounts in this chain to find the token
+            for (const accountTokens of Object.values(chainTokens)) {
+              if (Array.isArray(accountTokens)) {
+                const foundToken = accountTokens.find(
+                  (token: { address: string; decimals: number }) =>
+                    token.address.toLowerCase() === tokenAddress.toLowerCase(),
+                );
+                if (foundToken && foundToken.decimals !== undefined) {
+                  tokenDecimals = foundToken.decimals;
+                  break;
+                }
+              }
+            }
+          }
```

Comment:
Makes sense, just a tad gnarly to navigate through. Wondering if it is worth converting this into a indexible map object selector.